src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -799,20 +796,33 @@ DeferredErrorIfUnsupportedRecurringTuplesJoin(
 				continue;
 			}
 
-			if (ShouldRecurseForRecurringTuplesJoinChecks(outerrel) &&
-				RelationInfoContainsRecurringTuples(plannerInfo, outerrel, &recurType))
+
+			/*
+			 * If the outer side of the join doesn't have any distributed tables
+			 * (e.g., contains only recurring tuples), Citus should not pushdown
+			 * the query. The reason is that recurring tuples on every shard would
+			 * be added to the result, which is wrong.
+			 */
+			if (RelationInfoContainsOnlyRecurringTuples(plannerInfo, outerrel))
 			{
+				recurType = FetchFirstRecurType(plannerInfo, outerrel);
```

Comment:
Could FetchFirstRecurType and RelationInfoContainsOnlyRecurringTuples be merged?

Reply:
Because `RelationInfoContainsOnlyRecurringTuples ()` does not recurse to find `recurring tuples` but checks for existence of the distributed tables.

Also, we only need `FetchFirstRecurType ()` on failures, that's why I didn't care about more traversals.

Reply:
Ah, the difference between RelationInfoContainsRecurringTuples and RelationInfoContainsOnlyRecurringTuples was visually quite confusing

Reply:
Maybe add some comment on what we're doing here, e.g.
```c
/* find the first (or only) recurring RTE to give a meaningful error to the user */
```